session/types.proto

```diff
@@ -232,3 +278,114 @@ message ResponseVerificationHeader {
+    // NNS name is a domain name that resolves to an OwnerID through the
+    // NeoFS Name Service. The name must be a valid DNS-like domain name
+    // (e.g., "example.neofs") that is registered in the NNS contract on
+    // the Neo blockchain. The NNS record should contain a string record with
+    // the corresponding OwnerID value in NEP-18 address format.
+    string nns_name = 2 [json_name = "nnsName"];
+  }
+}
+
+// Lifetime parameters of the token v2. Field names taken from rfc7519.
+message TokenLifetime {
+  // Expiration, the last valid Unix timestamp.
+  uint64 exp = 1 [json_name = "exp"];
+
+  // Not valid before, the first valid Unix timestamp.
+  uint64 nbf = 2 [json_name = "nbf"];
+
+  // Issued at, the Unix timestamp when the token was issued.
+  uint64 iat = 3 [json_name = "iat"];
+}
+
+
+// DelegationInfo represents a single delegation in a chain of trust.
+message DelegationInfo {
+  // Account that performed this delegation.
+  Target issuer = 1 [json_name = "issuer"];
+
+  // Accounts that received the delegation.
+  // Maximum number of subjects: 100.
+  repeated Target subjects = 2 [json_name = "subjects"];
+
+  // Lifetime of this delegation.
+  // Allows each delegation in the chain to have its own validity window.
+  TokenLifetime lifetime = 3 [json_name = "lifetime"];
+
+  // List of verbs authorized by this delegation.
+  repeated Verb verbs = 4 [json_name = "verbs"];
+
+  // Signature of the issuer confirming this delegation record.
+  // The signature is created over the deterministic serialization
+  // of this DelegationInfo message excluding this field.
+  neo.fs.v2.refs.Signature signature = 5 [json_name = "signature"];
+}
+
+// SessionContextV2 carries unified context for both ObjectService and ContainerService requests.
+message SessionContextV2 {
+  // Container where operation is allowed.
+  // For container operations, this is the container being operated on.
+  // For object operations, this is the container holding the objects.
+  neo.fs.v2.refs.ContainerID container = 1 [json_name = "container"];
+
+  // Specific objects where operation is allowed.
+  // Only relevant for object operations.
+  // Empty list means all objects in the container.
+  // Maximum number of objects: 1000.
+  repeated neo.fs.v2.refs.ObjectID objects = 2 [json_name = "objects"];
+
+  // Operations authorized for this context.
+  repeated Verb verbs = 3 [json_name = "verbs"];
+}
+
+// SessionTokenV2 represents NeoFS Session Token with delegation support.
+message SessionTokenV2 {
+  // Session Token body.
+  message Body {
+    // Token version.
+    uint32 version = 1 [json_name = "version"];
+
+    // Token identifier (UUIDv4 in binary form).
+    bytes id = 2 [json_name = "id"];
+
+    // Account that issued this token (who signed it).
+    Target issuer = 3 [json_name = "issuer"];
+
+    // Accounts authorized by this token (who can use it).
+    // Maximum number of subjects: 100.
```

Comment:
8 for now, given that we can have proper groups this shouldn't be used for hundreds of things.